packages/kurt-vertex-ai/src/KurtVertexAI.ts

```diff
@@ -265,31 +265,60 @@ async function* transformStreamWithOptionalTools<
     for (const [partIndex, part] of parts.entries()) {
       const chunk = part.text
       const isFinal =
-        (isContentFinal && partIndex === parts.length - 1) || part.functionCall
+        (isContentFinal || part.functionCall) && partIndex === parts.length - 1
 
       if (chunk) {
         chunks.push(chunk)
         yield { chunk }
       }
       if (isFinal) {
-        const { functionCall } = part
-        if (functionCall) {
-          const { name } = functionCall
-          const schema = tools[name]
-          if (!schema) {
-            throw new Error(
-              `Vertex AI tried to call tool ${name} which isn't in the tool set ${JSON.stringify(
-                Object.keys(tools)
-              )}}`
-            )
+        if (part.functionCall) {
+          const allData = parts.map((part) => {
+            if (!part.functionCall) {
+              throw new Error(
+                `Vertex AI mixed function calls with non-function calls in the same raw stream event: ${JSON.stringify(
+                  rawEvent
+                )}`
+              )
+            }
+
+            const { name } = part.functionCall
+
+            const schema = tools[name]
+            if (!schema) {
+              throw new Error(
+                `Vertex AI tried to call tool ${name} which isn't in the tool set ${JSON.stringify(
+                  Object.keys(tools)
+                )}}`
+              )
+            }
+            return {
+              name,
+              args: applySchemaToFuzzyStructure(schema, part.functionCall),
+            } as D
+          })
+
+          // Emit a text chunk for each tool call (with line breaks in between).
+          for (const [dataIndex, data] of allData.entries()) {
+            if (dataIndex > 0) {
+              chunks.push("\n")
+              yield { chunk: "\n" }
+            }
+            const text = JSON.stringify(data.args)
+            chunks.push(text)
+            yield { chunk: text }
+          }
+
+          // biome-ignore lint/style/noNonNullAssertion: we already validated above that length > 0
```

Comment:
Where did you validated the `length > 0`?

Reply:
We're inside a loop of `parts` (at line 265 see `for (const [partIndex, part] of parts.entries()) { ...`), so we know that there is a nonzero number of `parts`, the `map` loop that generates the `allData` will give us an array with the same length as the `parts` array.

Reply:
Gotcha! Typescript is not smart enough to figure this one out, I think that I would rather we be explicit about this being an invariant when we create the `allData` array, rather than using the non null assertion.

```
const allData = const allData = parts.map(applySchemaToFuncionCall);
if(allData.length < 1) throw new Error("Encountered impossible state");

```
With this typescript should do a good job and expect allData[0] to be defined.

I like this approach better because if by any chance to stop checking, the non null assertion will continue to be there, and it will be hiding bugs.

Reply:
TypeScript still doesn't get it right, even with the length check -> throw happening first. Both the array index zero and the destructuring case still end up with type `D | undefined`:

```ts
          if (allData.length < 1)
            throw new Error("Empty here is impossible but TS doesn't know it")
          const [data, ...additionalData] = allData
          const data2 = allData[0]
```


<img width="614" alt="Screenshot 2024-06-03 at 7 51 09 AM" src="https://github.com/FormulaMonks/kurt/assets/4623870/bbfc9059-3b8e-4660-9f9a-3140aa62bfb2">

<img width="629" alt="Screenshot 2024-06-03 at 7 52 41 AM" src="https://github.com/FormulaMonks/kurt/assets/4623870/56809da6-e83f-480a-bf3b-2d961537e496">

Reply:
Okay, here's a solution that works:

```ts
          if (!arrayHasAtLeastOneElement(allData))
            throw new Error("Empty here is impossible but TS doesn't know it")
          const [data, ...additionalData] = allData
          const data2 = allData[0]
```

Where the `arrayHasAtLeastOneElement` function is defined as:

```ts
function arrayHasAtLeastOneElement<T>(array: T[]): array is [T, ...T[]] {
  return array.length > 0
}
```

<img width="621" alt="Screenshot 2024-06-03 at 7 59 44 AM" src="https://github.com/FormulaMonks/kurt/assets/4623870/48d128a6-ee9f-410e-9229-bf4bdf54fb09">
<img width="626" alt="Screenshot 2024-06-03 at 8 00 03 AM" src="https://github.com/FormulaMonks/kurt/assets/4623870/5fb9baba-8392-414e-a647-aa25fc39ac31">

Reply:
Haha, typescript can be a pain sometimes, yes, I agree with that solution. I see this generally named as `isNonEmptyArray` instead, so I would propose to name the function like that.